easyunfold/procar.py

```diff
@@ -32,85 +38,198 @@ def __init__(self, fobj_or_path=None, is_soc=False):
+
+        # normalise: (for each nspin, nkpt, nband, the sum of the projections over nion and proj_names should be 1)
+        proj_sum = np.sum(proj_data, axis=(-2, -1), keepdims=True)
+        proj_sum[proj_sum == 0] = 1  # just in case, avoid division by zero
+        proj_data /= proj_sum
+
+        if proj_xyz is not None:
+            proj_sum = np.sum(proj_xyz, axis=(-3, -2, -1), keepdims=True)
+            proj_sum[proj_sum == 0] = 1
+            proj_xyz /= proj_sum
+
+        return self.nspins, occs, kvecs, kweights, eigenvalues, proj_data, proj_xyz, parsed_kpoints
+
+    def _read_header_nion_proj_names(self, fobj):
+        """Read the header, nion and proj_names from the PROCAR"""
+        fobj.seek(0)
+        self.header = fobj.readline()
+        # Read the NK, NB and NIONS that are integers
+        _nkpts, _nbands, self.nion = [int(token) for token in re.sub(r'[^0-9]', ' ', fobj.readline()).split()]
+        self.proj_names = None  # projection names
+
+        for line in fobj:
+            if re.match(r'^ion.*tot', line):  # only the first "ion" line, in case of LORBIT >= 12
+                self.proj_names = line.strip().split()[1:-1]
+                break
+
+    def read(self, fobjs_or_paths):
+        """Read and amalgamate the data from a list of PROCARs"""
+
+        def open_file(fobj_or_path):
+            if isinstance(fobj_or_path, (str, Path)):
+                return open(fobj_or_path, encoding='utf-8')  # closed later
+            return fobj_or_path  # already a file-like object, just return it
+
+        parsed_kpoints = None
+        occs_list, kvecs_list, kweights_list = [], [], []
+        eigenvalues_list, proj_data_list, proj_xyz_list = [], [], []
+        for i, fobj_or_path in enumerate(fobjs_or_paths):
+            # Note: If PROCAR parsing becomes a significant bottleneck for people (e.g. with several HSE06+SOC PROCARs),
+            # this could be parallelized (somewhat) with multiprocessing. The actual file parsing in _read() is currently
+            # serial so no easy wins there, but could at least parallelise over the list of PROCARs
+            fobj = open_file(fobj_or_path)
+            if self.header is None:  # first file; read header, nion, proj_names
+                self._read_header_nion_proj_names(fobj)
+
+            current_nspins = self.nspins  # check spin consistency between PROCARs
+            nspins, occs, kvecs, kweights, eigenvalues, proj_data, proj_xyz, parsed_kpoints = self._read(fobj,
+                                                                                                         parsed_kpoints=parsed_kpoints)
+            if current_nspins is not None and current_nspins != nspins:
+                raise ValueError(f'Mismatch in number of spins in PROCARs supplied: ({nspins} vs {current_nspins})!')
+
+            if isinstance(fobj_or_path, (str, Path)):
+                fobj.close()  # if file was opened in this loop, close it
```

Comment:
```suggestion
            if opened:
                fobj.close()  # if file was opened in this loop, close it
```
Something like this.

Reply:
I think the current code is fine - just that there is no need to use `with` inside `_read` method.